ml_peg/analysis/utils/decorators.py

```diff
@@ -356,3 +356,189 @@ def build_table_wrapper(*args, **kwargs) -> dict[str, Any]:
+                Arguments to pass to the function being wrapped.
+            **kwargs
+                Key word arguments to pass to the function being wrapped.
+
+            Returns
+            -------
+            dict
+                Results dictionary.
+            """
+            results = func(*args, **kwargs)
+            # Form of results is
+            # results = {
+            #     metric_1: {mlip_1: value_1, mlip_2: value_2},
+            #     metric_2: {mlip_1: value_3, mlip_2: value_4},
+            # }
+
+            metrics_columns = ("MLIP",) + tuple(results.keys())
+            # Use MLIP keys from first (any) metric keys
+            mlips = next(iter(results.values())).keys()
+
+            metrics_data = []
+            for mlip in mlips:
+                metrics_data.append(
+                    {"MLIP": mlip}
+                    | {key: value[mlip] for key, value in results.items()}
+                    | {"id": mlip},
+                )
+
+            summary_tooltips = {
+                "MLIP": "Name of the model",
+                "Score": "Average of normalized metrics (higher is better)",
+                "Rank": "Model rank based on score (lower is better)",
+            }
+            if metric_tooltips:
+                tooltip_header = metric_tooltips | summary_tooltips
+            else:
+                tooltip_header = summary_tooltips
+
+            # Set default normalization ranges if not provided
+            default_ranges = {}
+            if normalization_ranges:
+                default_ranges = normalization_ranges.copy()
+
+            # Create default ranges for any missing metrics
+            for metric_name in results.keys():
+                if metric_name not in default_ranges:
+                    # Default: assume lower is better, set reasonable defaults
+                    values = [
+                        results[metric_name][mlip]
+                        for mlip in mlips
+                        if not np.isnan(results[metric_name][mlip])
+                    ]
+                    if values:
+                        min_val, max_val = min(values), max(values)
+                        # X (upper threshold) = min value (best case)
+                        # Y (lower threshold) = max value (worst case)
+                        default_ranges[metric_name] = (min_val, max_val)
+                    else:
+                        default_ranges[metric_name] = (0.0, 1.0)
+
+            # Apply normalization and calculate scores
+            # Calculate scores with provided or default normalizer
+            from ml_peg.analysis.utils import utils as _utils
+            from ml_peg.analysis.utils.utils import calc_ranks
+
+            if normalizer is None:
+                _normalizer = _utils.normalize_metric
+            else:
+                _normalizer = normalizer
+
+            # Local scorer using the chosen normalizer
+            def _calc_normalized_scores(rows, ranges):
```

Comment:
Is this not something we want as a general utility function that will be called later?